core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -226,50 +249,32 @@ private[spark] class ExecutorAllocationManager(
    * If the cap on the number of executors is reached, give up and reset the
    * number of executors to add next round instead of continuing to double it.
    * Return the number actually requested.
+   *
+   * @param maxNumNeededExecutors the maximum number of executors all currently running or pending
+   *                              tasks could fill
    */
-  private def addExecutors(): Int = synchronized {
+  private def addExecutors(maxNumNeededExecutors: Int): Int = synchronized {
     // Do not request more executors if we have already reached the upper bound
-    val numExistingExecutors = executorIds.size + numExecutorsPending
-    if (numExistingExecutors >= maxNumExecutors) {
+    val currentTarget = currentTargetNumExecutors
+    if (currentTarget >= maxNumExecutors) {
       logDebug(s"Not adding executors because there are already ${executorIds.size} " +
         s"registered and $numExecutorsPending pending executor(s) (limit $maxNumExecutors)")
       numExecutorsToAdd = 1
       return 0
     }
 
-    // The number of executors needed to satisfy all pending tasks is the number of tasks pending
-    // divided by the number of tasks each executor can fit, rounded up.
-    val maxNumExecutorsPending =
-      (listener.totalPendingTasks() + tasksPerExecutor - 1) / tasksPerExecutor
-    if (numExecutorsPending >= maxNumExecutorsPending) {
-      logDebug(s"Not adding executors because there are already $numExecutorsPending " +
-        s"pending and pending tasks could only fill $maxNumExecutorsPending")
-      numExecutorsToAdd = 1
-      return 0
-    }
-
-    // It's never useful to request more executors than could satisfy all the pending tasks, so
-    // cap request at that amount.
-    // Also cap request with respect to the configured upper bound.
-    val maxNumExecutorsToAdd = math.min(
-      maxNumExecutorsPending - numExecutorsPending,
-      maxNumExecutors - numExistingExecutors)
-    assert(maxNumExecutorsToAdd > 0)
-
-    val actualNumExecutorsToAdd = math.min(numExecutorsToAdd, maxNumExecutorsToAdd)
-
-    val newTotalExecutors = numExistingExecutors + actualNumExecutorsToAdd
-    val addRequestAcknowledged = testing || client.requestExecutors(actualNumExecutorsToAdd)
+    val newTotalExecutors = math.min(currentTarget + numExecutorsToAdd,
+      math.min(maxNumExecutors, maxNumNeededExecutors))
+    val numRequested = newTotalExecutors - currentTarget
```

Comment:
I think it's confusing to have both `numRequested` and `numExecutorsToAdd`. I prefer the old name `actualNumExecutorsToAdd` actually, and there we can add a short comment to explain what constitutes as "actual"